### DISCUSSION TITLE
Is there a way to statically check kernel arg types for their safety

### DISCUSSION THREAD

**Discussion by pauleonix** [NVIDIA/cccl#562] [Thrust]
I am working on a existing CUDA C++ codebase that combines Thrust fancy iterators with [eyalroz/cuda-api-wrappers](https://github.com/eyalroz/cuda-api-wrappers). In recent versions these wrappers have started statically checking the (decayed) type of kernel arguments passed through it's API for `std::is_trivially_copyable` or [even more recently](https://github.com/eyalroz/cuda-api-wrappers/commit/dfbeb0c0311c7d221a0018fb5336a85d2fc762c6)  `std::is_trivially_copy_constructible`. Neither of these requirements is met by Thrust fancy iterators like even the most basic `thrust::counting_iterator<int>`. As the two of us couldn't reach a conclusion or our own in https://github.com/eyalroz/cuda-api-wrappers/issues/551, we decided to ask here if there is another option to check for correctness without compromising composability with Thrust fancy iterators.

**Comment by miscco** [comment]
Thanks a lot, that is a very good question. 

Will discuss this in the coming week in the internal meeting. I need to have a look at our fancy iterator implementation though, because it is a bit dated

**Comment by eyalroz** [comment]
On my part, I definitely want to help resolve this, so please feel free to contact me if you want to discuss ideas which involve changes on both "sides", i.e. in my wrappers as well. My GitHub profile has my email, for coordinating something like that.

**Comment by pauleonix** [comment]
The [relevant section](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#global-function-argument-processing) of the Programming Guide says:

> When a `__global__` function is launched from host code, each argument type is allowed to be non-trivially copyable or non-trivially-destructible, but the processing for such types does not follow the standard C++ model, as described below. User code must ensure that this workflow does not affect program correctness.

**Reply by eyalroz** [reply]
@pauleonix  : What this quote is saying is, that the CUDA compiler (typically NVCC) will not _check_ for trivial-copy-constructibility, but that if a parameter is in fact, rather than just in principle, not-trivially-constructible or not-trivially-destructible, then things will get messed up, but that it is the user's responsibility. I think NVIDIA kind of copped out of doing the right thing here (or if you like: Did things the "C way" rather than the "C++ way").

It also tells me that I should probably also check for trivial destructibility...

**Reply by pauleonix** [reply]
"then things will get messed up": No, things **might** get messed up if and only if the copy-constructor has side effects or somehow else relies on being called a certain number of times. It's a safety vs flexibility tradeoff.

**Reply by jrhemstad** [reply]
As @pauleonix says, mountains of valid code that works just fine would have broken if nvcc had started rejecting non-trivially copyable types in kernel args. 

While I agree that the Thrust fancy iterators should be trivially copyable, it's not something on our immediate roadmap to address. We'd appreciate any contributions towards addressing this!

**Reply by eyalroz** [reply]
@pauleonix : Perhaps you want to make those iterators copy-constructible?

-- snip --

I was about to say something unkind here, but decided not to.

**Reply by pauleonix** [reply]
@jrhemstad I honestly [couldn't figure out](https://github.com/eyalroz/cuda-api-wrappers/issues/551#issuecomment-1762128987) what stops them from being trivially constructible last time I looked. Any pointers?

**Reply by miscco** [reply]
They are not trivially copy constructible, because they have user defined copy constructors. 

The issue is that we do not have the capacity to clean up all of thrust after dropping pre C++11 support.

**Reply by eyalroz** [reply]
> after dropping pre C++11 support.

isn't it easier to do cleanup after dropping legacy support?

Anyway, if the ctor's are effectively trivial, perhaps we could discuss a trait (within a detail namespace) for being either really-trivially-copy-constructible, or effectivelly-trivially-copy-constructible, i.e. so that a copy ctor may exist, but could be ignored when copying to the GPU in favor of trivial copy-construction. Then you would specialize that trait for relevant thrust ctors, and we'll be set.

**Reply by miscco** [reply]
Turns out that most of it was indeed relatively easy, so I opened a PR for it.

The issue is not whether it is easy to do something, but whether we actually have time for it. Today was a super fractured day for me, so I could take a bit of time to do something like this, but usually that is not the case.

**Comment by eyalroz** [comment]
> things **might** get messed up if and only if the copy-constructor has side effects or somehow else relies on being called a certain number of times.

That's what I meant by being not-trivially-copy-constructible "in fact, rather than just in principle"...

**Comment by eyalroz** [comment]
ping.

**Comment by miscco** [comment]
I have opened a PR that should address most of the fancy iterators bar `transform_iterator` https://github.com/NVIDIA/cccl/pull/1368

**Reply by eyalroz** [reply]
Wait, aren't you that journalist who testified before congress a while ago?

**Reply by miscco** [reply]
Oh boy, you know how many right wing bots I get on twitter for having a similar name?
